File: AlgorithmStudy/src/y2022/jul/Solution0711.java
package y2022.jul;

import model.TreeNode;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

// https://leetcode.com/problems/binary-tree-right-side-view/
public class Solution0711 {

    public List<Integer> rightSideView(TreeNode root) {
        /*
        트리를 오른쪽에서 봤을때 모습을 depth 순서대로 반환

        BFS -> 다 봐야 하고
            1      <-
           2  3    <-
         4      5  <-
           6       <-
           -> 1, 3, 5, 6

           q = [1]
             = [2,3]
             = [4,5]
             = [6]
        DFS -> [1,3,5,6] -> 다봐야해요.
         */
        List<Integer> ans = new ArrayList<>();

        if (root == null) {
            return ans;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);


        while (!queue.isEmpty()) {
            int size = queue.size();

            // depth
            TreeNode node = null;
            for (int i = 0; i < size; i++) {
                node = queue.poll();
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }
            ans.add(node.val);
        }

        return ans;

    }
}

File: AlgorithmStudy/src/y2023/jul/Solution0711.kt
package y2023.jul

import model.TreeNode
import java.util.*

// https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/
class Solution0711 {
    fun distanceK(root: TreeNode?, target: TreeNode?, k: Int): List<Int> {
        if (root == null || target == null) {
            return listOf()
        }
        /*if ( k == 0 ) {
            return listOf(target.`val`)
        }*/
        // target Node 먼저 구하고
        val res = mutableListOf<Int>()
        val parentMap = mutableMapOf<TreeNode, TreeNode>()
        val visitedMap = mutableMapOf<TreeNode, Boolean>()
        val queue = LinkedList<TreeNode>()
        queue.offer(root)

        while(queue.isNotEmpty()) {
            val node = queue.poll()
            visitedMap[node] = false
            node.left?.let {
                parentMap[it] = node
                queue.offer(it)
            }
            node.right?.let {
                parentMap[it] = node
                queue.offer(it)
            }
        }

        queue.offer(target)
        visitedMap[target] = true
        var depth = 0
        while(queue.isNotEmpty()) {
            val size = queue.size

            if (depth == k) {
                while(queue.isNotEmpty()) {
                    res.add(queue.poll().`val`)
                }
                break
            }
            for(i in 0 until size) {
                val node = queue.poll()
                visitedMap[node] = true

                node.left?.let {
                    if(visitedMap[it] == false)
                        queue.offer(it)
                }

                node.right?.let {
                    if (visitedMap[it] == false) {
                        queue.offer(it)
                    }
                }

                parentMap[node]?.let {
                    if (visitedMap[it] == false) {
                        queue.offer(it)
                    }
                }
            }
            depth++

        }

        return res
    }
}
